### PR TITLE
4.x - PSR-18

### PR DESCRIPTION
**Pull request by ADmad**
I'll add more tests and exception handling as per PSR-18 specs if the updates till now seem fine.

### CODE REVIEW

File: src/Http/Client.php [hunk 1]
```diff
@@ -396,17 +399,29 @@ protected function _mergeOptions(array $options): array
...
+    /**
+     * Sends a PSR-7 request and returns a PSR-7 response.
+     *
+     * @param \Psr\Http\Message\RequestInterface $request Request instance.
+     * @return \Psr\Http\Message\ResponseInterface Response instance.
+     * @throws \Psr\Http\Client\ClientExceptionInterface If an error happens while processing the request.
+     */
+    public function sendRequest(RequestInterface $request): ResponseInterface
+    {
+        return $response = $this->send($request, $this->_config);
```
Comment (saeideng) [inline src/Http/Client.php]: redundant?? `$response =`

Comment (ADmad) [reply src/Http/Client.php]: Yup, I'll fix it.

**Review by markstory — APPROVED** [review]
Looks good to me. Nice that this was so simple to do :clap:

**Comment by ADmad** [comment]
The static analysis job failure is unrelated.

File: src/Http/Client/Adapter/Stream.php [hunk 2]
```diff
@@ -268,7 +269,7 @@ protected function _send(RequestInterface $request): array
         fclose($this->_stream);
 
         if ($timedOut) {
-            throw new HttpException('Connection timed out ' . $url, 504);
+            throw new NetworkException('Connection timed out ' . $url, $request);
```
Comment (markstory) [inline src/Http/Client/Adapter/Stream.php]: This change will break the behavior where an http request to a backend timing out will result in the app also responding with a 504.

Comment (ADmad) [reply src/Http/Client/Adapter/Stream.php]: I can make `NetworkException` extend `HttpException` and set `504` status code but personally I don't think a timeout on an internal request should bubble up as `504` for the parent request. It should still be a 500 IMO.

Comment (ADmad) [reply src/Http/Client/Adapter/Stream.php]: I checked a couple of guzzle wrappers for PSR18 support and couple of standalone PSR18 implementations and non set a `504` code.

Comment (markstory) [reply src/Http/Client/Adapter/Stream.php]: Ok. Then perhaps our code should change too.

**Comment by ADmad** [comment]
This is ready for final review. I have also added `ClientException` which is thrown when the adapter can't even initiate a request if required PHP config/module is disabled.

**Comment by markstory** [comment]
Looks good other than builds failing.

**Pull request merged**